Buildx by default suggestions

thaJeztah opened the pull request:
Relates to https://github.com/docker/cli/pull/3314

Thought it'd be easier this way; input/suggestions welcome (also feel free to squash etc)

crazy-max committed:
Set buildx as default builder

Signed-off-by: CrazyMax <crazy-max@users.noreply.github.com>

crazy-max committed:
Fallback to legacy builder

Signed-off-by: CrazyMax <crazy-max@users.noreply.github.com>

thaJeztah committed:
builder: simplify error generation, and rephrase error/warning

With this change:

    echo 'FROM busybox' | DOCKER_BUILDKIT=1 docker build -
    ERROR: BuildKit is enabled but the buildx component is missing or broken.
           Install the buildx component to build images with BuildKit:
           https://docs.docker.com/go/buildx/

    echo 'FROM busybox' | docker build -
    DEPRECATED: The legacy builder is deprecated and will be removed in a future release.
                Install the buildx component to build images with BuildKit:
                https://docs.docker.com/go/buildx/

    Sending build context to Docker daemon  2.048kB
    ...

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
Update deprecation docs

- expand a bit on what's happening
- clarify start of deprecation of the classic builder
- show examples of error and warning

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah commented on e2e/image/build_test.go:
```diff
@@ -37,7 +37,12 @@ func TestBuildFromContextDirectoryWithTag(t *testing.T) {
 	defer icmd.RunCommand("docker", "image", "rm", "myimage")
 
-	result.Assert(t, icmd.Expected{Err: "DEPRECATED: The legacy builder is deprecated and will be removed in a future release."})
+	const buildxMissingWarning = `DEPRECATED: The legacy builder is deprecated and will be removed in a future release.
```
Haven't run the e2e tests, so they may fail here (whitespacing and such)

crazy-max replied:
Yes looks like this change can be reverted now that the warning message has been removed from build pkg.

```
=== FAIL: e2e/image TestBuildFromContextDirectoryWithTag (2.81s)
    build_test.go:45: assertion failed: 
        Command:  docker build -t myimage .
        ExitCode: 0
        Stdout:   Sending build context to Docker daemon  4.608kB

        Step 1/4 : FROM registry:5000/alpine:3.6
         ---> 76da55c8019d
        Step 2/4 : COPY run /usr/bin/run
         ---> 485496f425b9
        Step 3/4 : RUN run
         ---> Running in 48dd9744aa87
        running
        Removing intermediate container 48dd9744aa87
         ---> 076deca8990c
        Step 4/4 : COPY data /data
         ---> 0ee25cd5a8cb
        Successfully built 0ee25cd5a8cb
        Successfully tagged myimage:latest
        
        Stderr:   
        
        Failures:
        Expected stderr to contain "DEPRECATED: The legacy builder is deprecated and will be removed in a future release.\n            Install the buildx component to build images with BuildKit:\n            https://docs.docker.com/go/buildx/\n"
```

But in this test it should display a warning about the deprecated legacy builder and it doesn't seem to pop out.

crazy-max commented on cmd/docker/builder.go:
```diff
@@ -13,44 +12,29 @@ import (
-}
+	builderDefaultPlugin = "buildx"
+	buildxMissingWarning = `DEPRECATED: The legacy builder is deprecated and will be removed in a future release.
+            Install the buildx component to build images with BuildKit:
```
```suggestion
            Install the %s component to build images with BuildKit:
```

We have to specify the builder name dynamically in case a user use the non-default (buildx) one.

thaJeztah replied:
You mean `buildx` without the `docker-` prefix? Isn't that a "incorrect" installation (i.e.; rename the binary)?

crazy-max replied:
If someone has an alias builder in his configuration we should use it. So if the alias builder is missing or broken that's the one we should display in the warning/error msg.

thaJeztah replied:
So; 

- the `aliases` option was added specifically for buildx (to allow it to override the built-in `docker build` (https://github.com/docker/cli/pull/1840), and [not documented](https://docs.docker.com/engine/reference/commandline/cli/#cli-plugin-options) - deliberately as it was considered a temporary "hack" at the time. It was not the intent to allow `docker build` to be replaced with `<insert any other plugin>`
- even _if_ the aliases option would be kept, we'd still be only interested in `buildx`, not in any other plugin that's configured to replace `docker build` I don't think it makes sense to provide this error for `<insert some plugin>` (install the `kaniko` component to build images with BuildKit?) (the intent of the `buildx by default` change is to provide a consistent experience; 
- because of the above, I think we should consider removing the aliases option altogether, as the reason it was added (replace the built-in BuildKit client with the buildx one) is now the default, so we no longer need the option. (that is: remove the configuration option (or ignore it); if needed, we can keep the code logic in case other commands would be moved out of the main CLI (although there's no plans for that))

tonistiigi replied:
This is just internal of how the plugins are loaded. We don't want to confuse users with this. The text should just say "buildx".

crazy-max replied:
> * the `aliases` option was added specifically for buildx (to allow it to override the built-in `docker build` (

Ah ok that makes sense, thanks!

crazy-max replied:
Do you think we should deprecate this `aliases` in the future?

tonistiigi replied:
> Do you think we should deprecate the aliases in the future?

It's fine as a way to organize things and leaves more options for the future. But users should not worry about these internals.

thaJeztah commented on cli/command/image/build.go:
```diff
@@ -184,12 +184,6 @@ func runBuild(dockerCli command.Cli, options buildOptions) error {
 		remote        string
 	)
 
-	if !options.quiet && dockerCli.ServerInfo().OSType != "windows" {
```
Reminds me; I didn't try if the other location takes `options.quiet` into account

crazy-max replied:
No it's only taken into account in build pkg that's why it was easier to add the message here. We would need to parse the command and check if quiet is enabled and a build command is being used. I don't think we should do that their imo.

crazy-max commented on docs/deprecated.md:
```diff
@@ -100,20 +101,82 @@ Removed    | [`--api-enable-cors` flag on `dockerd`](#--api-enable-cors-flag-on-
 Removed    | [`--run` flag on `docker commit`](#--run-flag-on-docker-commit)                                                                    | v0.10      | v1.13
 Removed    | [Three arguments form in `docker import`](#three-arguments-form-in-docker-import)                                                  | v0.6.7     | v1.12
 
+### Legacy builder for Linux images
+
+**Deprecated in Release: v21.xx**
+
+Docker v21.xx now uses BuildKit by default to build Linux images, and uses the
+[Buildx](https://docs.docker.com/buildx/working-with-buildx/) CLI component for
+`docker build`. With this change, `docker build` now exposes all advanced features
+that BuildKit provides and which were previously only available through the
+`docker buildx` subcommands.
+
+The Buildx component is installed automatically when installing the `docker` CLI
+using our `.deb` or `.rpm` packages, and statically linked binaries are provided
+both on `download.docker.com`, and through the [`docker/buildx-bin` image](https://hub.docker.com/r/docker/buildx-bin)
+on Docker Hub. Refer the [Buildx section](http://docs.docker.com/go/buildx/) for
+detailed instructions on installing the Buildx component.
+
+This release marks the beginning of the deprecation cycle of the classic ("legacy")
+builder for Linux images. No active development will happen on the classic builder
+(except for bugfixes). BuildKit development started five Years ago, left the
+"experimental" phase since Docker 18.09, and is already the default builder for
+[Docker Desktop](https://docs.docker.com/desktop/mac/release-notes/3.x/#docker-desktop-320).
+While we're comfortable that BuildKit is stable for general use, there may be
+some changes in behavior. If you encounter issues with BuildKit, we encourage
+you to report issues in the [BuildKit issue tracker on GitHub](https://github.com/moby/buildkit/){:target="_blank" rel="noopener" class="_"}
+
+> Classic builder for building Windows images
+>
+> BuildKit does not (yet) provide support for building Windows images, and
+> `docker build` continues to use the classic builder to build native Windows
+> images on Windows daemons.
```
Sound great thanks!

crazy-max committed:
Fix install buildx uri

Signed-off-by: CrazyMax <crazy-max@users.noreply.github.com>

crazy-max commented:
Sorry I have to rebase since latest changes upstream :sweat_smile:

thaJeztah committed:
remove some of the alias logic

The aliases (configurable through the `~/.docker/config.json`) were a temporary
solution while buildx was not yet the default. They're an internal-use (undocumented)
option, so we don't have to take other uses into account.

This removes some of the logic to read the aliasses-config from the configuration
file, and hard-codes it to assume `buildx`.

(Probably more code can be removed/simplified)

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah commented:
@crazy-max I had some draft changes to remove the `aliases` logic https://github.com/crazy-max/docker-cli/pull/3#discussion_r794012482

pushed it as a commit (but it may need more cleaning up)
